dhall/src/Dhall/Eval.hs

```diff
@@ -0,0 +1,900 @@
+
+    convE :: Val a -> Val a -> Bool
+    convE = conv env
+    {-# inline convE #-}
+
+    convSkip :: Text -> Val a -> Val a -> Bool
+    convSkip x = conv (Skip env x)
+    {-# inline convSkip #-}
+
+  in case (t, t') of
+    (VConst k, VConst k') -> k == k'
+    (VVar x i, VVar x' i') -> x == x' && i == i'
+
+    (VLam _ (freshCl -> (x, v, t)), VLam _ t' ) -> convSkip x (inst t v) (inst t' v)
+    (VLam _ (freshCl -> (x, v, t)), VHLam _ t') -> convSkip x (inst t v) (t' v)
+    (VLam _ (freshCl -> (x, v, t)), t'        ) -> convSkip x (inst t v) (vApp t' v)
+    (VHLam _ t, VLam _ (freshCl -> (x, v, t'))) -> convSkip x (t v) (inst t' v)
+    (VHLam _ t, VHLam _ t'                    ) -> let (x, v) = fresh "x" in convSkip x (t v) (t' v)
+    (VHLam _ t, t'                            ) -> let (x, v) = fresh "x" in convSkip x (t v) (vApp t' v)
+
+    (t, VLam _ (freshCl -> (x, v, t'))) -> convSkip x (vApp t v) (inst t' v)
+    (t, VHLam _ t'  ) -> let (x, v) = fresh "x" in convSkip x (vApp t v) (t' v)
+
+    (VApp t u, VApp t' u') -> convE t t' && convE u u'
+
+    (VPi a b, VPi a' (freshCl -> (x, v, b'))) ->
+      convE a a' && convSkip x (inst b v) (inst b' v)
+    (VPi a b, VHPi (fresh -> (x, v)) a' b') ->
+      convE a a' && convSkip x (inst b v) (b' v)
+    (VHPi _ a b, VPi a' (freshCl -> (x, v, b'))) ->
+      convE a a' && convSkip x (b v) (inst b' v)
+    (VHPi _ a b, VHPi (fresh -> (x, v)) a' b') ->
+      convE a a' && convSkip x (b v) (b' v)
+
+    (VBool       , VBool            ) -> True
+    (VBoolLit b  , VBoolLit b'      ) -> b == b'
+    (VBoolAnd t u, VBoolAnd t' u'   ) -> convE t t' && convE u u'
+    (VBoolOr  t u, VBoolOr  t' u'   ) -> convE t t' && convE u u'
+    (VBoolEQ  t u, VBoolEQ  t' u'   ) -> convE t t' && convE u u'
+    (VBoolNE  t u, VBoolNE  t' u'   ) -> convE t t' && convE u u'
+    (VBoolIf t u v, VBoolIf t' u' v') -> convE t t' && convE u u' && convE v v'
+
+    (VNatural, VNatural) -> True
+    (VNaturalLit n, VNaturalLit n') -> n == n'
+    (VNaturalFold t _ u v, VNaturalFold t' _ u' v') ->
+      convE t t' && convE u u' && convE v v'
+
+    (VNaturalBuild t     , VNaturalBuild t')     -> convE t t'
+    (VNaturalIsZero t    , VNaturalIsZero t')    -> convE t t'
+    (VNaturalEven t      , VNaturalEven t')      -> convE t t'
+    (VNaturalOdd t       , VNaturalOdd t')       -> convE t t'
+    (VNaturalToInteger t , VNaturalToInteger t') -> convE t t'
+    (VNaturalShow t      , VNaturalShow t')      -> convE t t'
+    (VNaturalPlus t u    , VNaturalPlus t' u')   -> convE t t' && convE u u'
+    (VNaturalTimes t u   , VNaturalTimes t' u')  -> convE t t' && convE u u'
+
+    (VInteger           , VInteger)            -> True
+    (VIntegerLit t      , VIntegerLit t')      -> t == t'
+    (VIntegerShow t     , VIntegerShow t')     -> convE t t'
+    (VIntegerToDouble t , VIntegerToDouble t') -> convE t t'
+
+    (VDouble       , VDouble)        -> True
+    (VDoubleLit n  , VDoubleLit n')  -> n == n'
```

Comment:
OK. I think now I should just export a `doubleToTerm :: Double -> Term` function from `Dhall.Binary`, and use that (I don't want to use `encode` because of the unnecessary `ToTerm` constraint on `conv`; also, the next iteration of `conv` likely will not have any class constraint, because it'll have access to `Val`-s of imported expressions).

Reply:
Another way to do this without a `ToTerm` constraint is to wrap the `Double` value in a `DoubleLit` constructor and then `Dhall.Binary.encode` that

Reply:
Right, I realize now I just have to specify an arbitrary `Expr s a` type for the `DoubleLit`.

Reply:
I pushed a commit, I used `encode` as you suggested.